crates/sui-core/src/authority_aggregator.rs

```diff
@@ -1053,20 +1054,53 @@ where
         }
     }
 
-    fn handle_transaction_response_with_signed(
+    fn handle_transaction_response(
         &self,
         state: &mut ProcessTransactionState,
-        signed_transaction: VerifiedSignedTransaction,
+        plain_tx: SignedTransaction,
     ) -> SuiResult<Option<ProcessTransactionResult>> {
-        let (data, sig) = signed_transaction.into_inner().into_data_and_sig();
+        let (data, sig) = plain_tx.into_data_and_sig();
         match state.tx_signatures.insert(sig) {
             InsertResult::NotEnoughVotes => Ok(None),
             InsertResult::Failed { error } => Err(error),
             InsertResult::QuorumReached(cert_sig) => {
-                let ct = CertifiedTransaction::new_from_data_and_sig(data, cert_sig);
-                Ok(Some(ProcessTransactionResult::Certified(
-                    ct.verify(&self.committee)?,
-                )))
+                let ct = CertifiedTransaction::new_from_data_and_sig(data.clone(), cert_sig);
+                let res = ct.verify(&self.committee);
+                match res {
+                    // If the aggregated signature verifies ok, then returns the ceritified transaction.
+                    Ok(ct) => Ok(Some(ProcessTransactionResult::Certified(ct))),
+                    Err(_) => {
+                        debug!(
+                            "Fail to verify the aggregated siganture for tx: {:?}",
+                            data.digest()
+                        );
+                        // If the aggregated signature fails to verify, iterate through all signatures and verify individually.
+                        // Increment the bad stake and continue to find new stake to reach the quorum.
+                        //
+                        // TODO(joyqvq): It is possible for the aggregated signature to fail every time when the last one
+                        // single signauture fails to verify repeatedly, and trigger this for loop to run. This can be
+                        // optimized by caching single sigs that are already verified and only verify the net new ones.
+                        // Same for [fn handle_transaction_response_with_executed].
+                        for (name, s) in &state.tx_signatures.clone().data {
```

Comment:
I think you can make this a API inside the stake aggregator, that would also remove the bad signatures.

Reply:
in order to verify the sig in the stake aggregator, i need tx_signatures.insert to take in an envelope<T, AuthoritySignInfo> where T: Message + Serialize.